tests/ga/test_update.py

```diff
@@ -2110,6 +2110,141 @@ class ReportStatusTestCase(AgentTestCase):
+            patch("sys.exit")
+        ]
+
+        for p in self.patches:
+            p.start()
+
+        return AgentTestCase.setUp(self)
+
+    def tearDown(self):
+        
+        for p in self.patches:
+            p.stop()
+
+        return AgentTestCase.tearDown(self)
+
+    @staticmethod
+    @contextlib.contextmanager
+    def _mock_update_handler(iterations=1, **kwargs):
+        """
+        Creates an UpdateHandler instance that will run n iterations.
+
+        Can be supplied keyword args for:
+            *   mock_wire_data_file: This arg is treated like mock_wire_protocol
+                would. Defaults to mockwiredata.DATA_FILE_STATUS_BLOB
+            *   http_<action>_handler where action is get, put, or post: This arg
+                is treated like mock_wire_protocol would.
+        
+        Returned UpdateHandler instance has its protocol mocked via mock_wire_protocol
+        and its _report_status function wrapped in a mock for accounting purposes.
+        """
+        
+        # Build the side_effect list for the UpdateHandler.is_running PropertyMock.
+        # Return True for the first n iterations followed by a single False to stop
+        # and then another True because the current UpdateHanlder implementation 
+        # does a __set__ during shutdown.
+        is_running_return_values = [True] * iterations + [False, True]
+        is_running_patch = patch.object(UpdateHandler, "is_running", PropertyMock(side_effect=is_running_return_values))
+
+        mock_http_get = kwargs.get("http_get_handler")
+        mock_http_put = kwargs.get("http_put_handler")
+        mock_http_post = kwargs.get("http_post_handler")
+
+        mock_wire_data_file = kwargs.get("mock_wire_data_file", mockwiredata.DATA_FILE_STATUS_BLOB)
+
+        try:
+            with mock_wire_protocol(mock_wire_data_file, mock_http_get, mock_http_post, mock_http_put) as protocol:
+                update_handler = get_update_handler()
+                
+                update_handler.protocol_util.get_protocol = Mock(return_value=protocol)
+
+                original_report_status = update_handler._report_status
+                update_handler._report_status = Mock(side_effect=original_report_status)
+
+                is_running_patch.start()
+                yield update_handler
+        finally:
+            is_running_patch.stop()
+
+    @staticmethod
+    def _goal_state_handler(url, **kwargs): # pylint: disable=unused-argument
+        """
+        Note: this function's signature has to coincide with that of http_get_request,
+        and thus needs **kwargs. We don't need the information, though, so we tell pylint 
+        we aren't planning on using the argument.
+
+        Relies on this function to have the property return_vals, so it'll need to be
+        set before being called
+        """
+        if not HttpRequestPredicates.is_goal_state_request(url):
+            return None
```

Comment:
If its not a GoalState request, it should return success right?

Reply:
Yes, I can make it more explicit, but that is what this line is already doing. The `mock_wire_protocol` context manager basically queries the appropriate provided handler (i.e., this function is called on GET requests) and calls the default handler if the provided handler returns `None`.

Essentially, the way to get the mock protocol to return the file data is to return `None`.

Reply:
> Essentially, the way to get the mock protocol to return the file data is to return None.

Are you sure about this? From what I remember returning a `None` is an error. I mean that is exactly what you're doing too right when you set this mock - 

`ReportStatusTestCase._goal_state_handler.return_vals = [ None, None ]`

Reply:
Yes, the code in question related to this is `tests/protocol/mocks.py#85-103`:

```python
if handler is not None:
    if method == 'GET':
        return_value = handler(url, **kwargs)
    else:
        return_value = handler(url, data, **kwargs)
    if return_value is not None:
        tracked_urls.append(url)
        if isinstance(return_value, Exception):
            raise return_value
        return return_value
        
# if the request was not handled try to use the mock wireserver data
if _USE_MOCK_WIRE_DATA_RE.match(url) is not None:
    if method == 'GET':
        return protocol.mock_wire_data.mock_http_get(url, **kwargs)
    if method == 'POST':
        return protocol.mock_wire_data.mock_http_post(url, data, **kwargs)
    if method == 'PUT':
        return protocol.mock_wire_data.mock_http_put(url, data, **kwargs)
```

And this is documented in `mock_wire_protocol`'s docstring (last line in the quote is most important here):

```
The caller can also provide handler functions for specific HTTP methods using the http_*_handler arguments. The return value of the handler
function is interpreted similarly to the "return_value" argument of patch(): if it is an exception the exception is raised or, if it is
any object other than None, the value is returned by the mock. If the handler function returns None the call is handled using the mock
wireserver data or passed to the original to restutil.http_request.
```

What `ReportStatusTestCase._goal_state_handler.return_vals = [ None, None ]` is doing is determining how many times a goal state request succeeds before failing with a `HttpError` (via raisining an `indexError` when `...return_vals.pop()` is called on an empty array). I can add to the comment right above that line, mentioning that returning `None` has a special meaning to `mock_wire_protocol` that we inherit from `patch()`.

Reply:
Ahh I see now, it makes sense thank you. A comment would be super helpful here, I did not get this behavior at all at first glance, I thought that you were implementing some sort of a patch behavior where each `None` represented the value that will be returned when the code tries to do a GET.